File: R&D/elevenlabs/poc/server/src/utils/audio.ts
const MULAW_MAX = 255;
const SIGN_BIT = 0x80;
const QUANT_MASK = 0x0f;
const SEG_MASK = 0x70;
const SEG_SHIFT = 4;
const BIAS = 0x84;

function muLawDecodeSample(uVal: number): number {
  uVal = ~uVal;
  const sign = (uVal & SIGN_BIT) ? -1 : 1;
  let exponent = (uVal & SEG_MASK) >> SEG_SHIFT;
  let mantissa = uVal & QUANT_MASK;
  mantissa = mantissa + 0.5;
  const magnitude = ((mantissa << (exponent + 3)) + BIAS) - BIAS;
  return sign * magnitude;
}

function muLawEncodeSample(sample: number): number {
  let sign = (sample < 0) ? SIGN_BIT : 0;
  sample = Math.abs(sample);

  if (sample > 32635) {
    sample = 32635;
  }

  sample = sample + BIAS;

  let exponent = 7;
  for (let expMask = 0x4000; (sample & expMask) === 0 && exponent > 0; expMask >>= 1) {
    exponent--;
  }

  let mantissa = (sample >> (exponent + 3)) & QUANT_MASK;
  const muLawByte = ~(sign | (exponent << SEG_SHIFT) | mantissa);
  return muLawByte & MULAW_MAX;
}

export function decodeMuLawBase64ToPcm16(base64: string): Buffer {
  const mulaw = Buffer.from(base64, 'base64');
  const pcm = Buffer.alloc(mulaw.length * 2);

  for (let i = 0; i < mulaw.length; i += 1) {
    const decoded = muLawDecodeSample(mulaw[i]);
    pcm.writeInt16LE(decoded, i * 2);
  }

  return pcm;
}

export function encodePcm16ToMuLawBase64(pcm: Buffer): string {
  const samples = pcm.length / 2;
  const mulaw = Buffer.alloc(samples);

  for (let i = 0; i < samples; i += 1) {
    const sample = pcm.readInt16LE(i * 2);
    const encoded = muLawEncodeSample(sample);
    mulaw[i] = encoded;
  }

  return mulaw.toString('base64');
}

File: R&D/elevenlabs/poc/python-server/app/utils/audio.py
from __future__ import annotations

import base64
import math

_SIGN_BIT = 0x80
_QUANT_MASK = 0x0F
_SEG_SHIFT = 4
_SEG_MASK = 0x70
_BIAS = 0x84


def _search_segment(value: int) -> int:
    for segment in range(7):
        if value <= (0x1F << segment):
            return segment
    return 7


def encode_pcm16_to_mulaw_base64(pcm: bytes) -> str:
    samples = len(pcm) // 2
    mulaw = bytearray(samples)

    for i in range(samples):
        sample = int.from_bytes(pcm[i * 2 : i * 2 + 2], "little", signed=True)
        sign = _SIGN_BIT if sample < 0 else 0
        sample = abs(sample)

        if sample > 32635:
            sample = 32635

        sample += _BIAS
        segment = _search_segment(sample >> 7)
        mantissa = (sample >> (segment + 3)) & _QUANT_MASK

        encoded = ~(sign | (segment << _SEG_SHIFT) | mantissa) & 0xFF
        mulaw[i] = encoded

    return base64.b64encode(mulaw).decode("ascii")


def decode_mulaw_base64_to_pcm16(payload: str) -> bytes:
    mulaw = base64.b64decode(payload)
    pcm = bytearray(len(mulaw) * 2)

    for i, byte in enumerate(mulaw):
        byte = ~byte & 0xFF
        sign = -1 if (byte & _SIGN_BIT) else 1
        segment = (byte & _SEG_MASK) >> _SEG_SHIFT
        mantissa = byte & _QUANT_MASK
        magnitude = ((mantissa + 0.5) * (1 << (segment + 3))) - _BIAS
        sample = int(sign * magnitude)
        pcm[i * 2 : i * 2 + 2] = sample.to_bytes(2, "little", signed=True)

    return bytes(pcm)



